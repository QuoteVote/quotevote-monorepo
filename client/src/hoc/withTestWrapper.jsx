import React from 'react'
import { PersistGate } from 'redux-persist/integration/react'
import { ApolloProvider } from '@apollo/react-hooks'
import { Provider } from 'react-redux'
import { createTheme, ThemeProvider } from '@material-ui/core/styles'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import store, { persistor } from '../store/store'
import client from '../config/apollo'
import customTheme from '../theme'
import { AuthModalProvider } from '../Context/AuthModalContext'
const hist = createBrowserHistory()
const theme = createTheme(customTheme)

const withTestWrapper = (Component) => (props) => (
  <ApolloProvider client={client}>
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <ThemeProvider theme={theme}>
          <AuthModalProvider>
            <Router history={hist}>
              <Component {...props} />
            </Router>
          </AuthModalProvider>
        </ThemeProvider>
      </PersistGate>
    </Provider>
  </ApolloProvider>
)
export default withTestWrapper
