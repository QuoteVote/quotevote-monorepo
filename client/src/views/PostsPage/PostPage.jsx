import { useEffect, useState, useMemo } from 'react';
import { Helmet } from 'react-helmet-async';
import PropTypes from 'prop-types';
import { Grid, useMediaQuery } from '@material-ui/core';
import { makeStyles, useTheme } from '@material-ui/core/styles';
import { useQuery, useSubscription } from '@apollo/react-hooks';
import { useSelector, useDispatch } from 'react-redux';
import { isEmpty } from 'lodash';
import { Redirect } from 'react-router-dom';
import Post from '../../components/Post/Post';
import PostActionList from '../../components/PostActions/PostActionList';
import PostSkeleton from '../../components/Post/PostSkeleton';
import { GET_ROOM_MESSAGES, GET_POST } from '../../graphql/query';
import { NEW_MESSAGE_SUBSCRIPTION } from '../../graphql/subscription';
import PostChatSend from '../../components/PostChat/PostChatSend';
import { tokenValidator } from 'store/user';

const useStyles = makeStyles((theme) => ({
  root: {
    height: '85vh',
    overflow: 'hidden',
    marginTop: 0,
    width: '100% !important',
    marginLeft: '0 !important',
    marginRight: '0 !important',
    padding: '0 !important',
  },
  mobileContainer: {
    height: '85vh',
    maxHeight: '100vh',
    overflow: 'hidden',
    marginTop: 0,
    display: 'flex',
    flexDirection: 'column',
  },
  desktopContainer: {
    height: '85vh',
    maxHeight: '85vh',
    overflow: 'hidden',
    marginTop: 0,
    marginLeft: '0 !important',
    marginRight: '0 !important',
    padding: '0 !important',
    display: 'flex',
    width: '100% !important',
  },
  mobilePostSection: {
    flex: '0 0 auto',
    maxHeight: '40vh',
    overflow: 'auto',
    padding: theme.spacing(2),
  },
  desktopPostSection: {
    flex: '0 0 50%',
    height: '85vh',
    overflow: 'auto',
    padding: 0,
    borderRight: `1px solid ${theme.palette.divider}`,
  },
  mobileInteractionSection: {
    flex: 1,
    display: 'flex',
    flexDirection: 'column',
    minHeight: 0,
    position: 'relative',
  },
  mobileMessagesContainer: {
    flex: 1,
    overflow: 'auto',
    padding: theme.spacing(2),
    paddingBottom: theme.spacing(8),
  },
  mobileChatInputContainer: {
    position: 'fixed',
    bottom: 0,
    left: 0,
    right: 0,
    zIndex: 1000,
    backgroundColor: theme.palette.background.paper,
    borderTop: `1px solid ${theme.palette.divider}`,
    boxShadow: '0 -2px 10px rgba(0,0,0,0.1)',
  },
  desktopInteractionSection: {
    flex: '0 0 50%',
    height: '85vh',
    display: 'flex',
    flexDirection: 'column',
    overflow: 'hidden',
  },
  desktopMessagesContainer: {
    flex: 1,
    overflow: 'auto',
    padding: 0,
    paddingBottom: theme.spacing(1),
  },
  desktopChatInputContainer: {
    marginLeft: 0,
    flexShrink: 0,
    borderTop: `1px solid ${theme.palette.divider}`,
    backgroundColor: theme.palette.background.paper,
  },
  emptyPost: {
    marginTop: 100,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
  },
  link: {
    color: '#00bcd4',
  },
}))

function PostPage({ postId }) {
  const classes = useStyles()
  const theme = useTheme()
  const isMobile = useMediaQuery(theme.breakpoints.down('sm'))
  const [postHeight, setPostHeight] = useState()
  const dispatch = useDispatch()

  const idSelector = useSelector((state) => state.ui.selectedPost.id)
  const user = useSelector((state) => state.user.data)

  // Check if user is authenticated
  const isAuthenticated = user && user._id && tokenValidator(dispatch)

  const {
    loading: loadingPost,
    error: postError,
    data: postData,
    refetch: refetchPost,
  } = useQuery(GET_POST, {
    variables: { postId },
  })

  const { post } = (!loadingPost && postData) || {}

  // Open Graph/Twitter meta values
  const ogTitle = post?.title || 'Quote.Vote â€“ The Internet\'s Quote Board';
  const ogDescription = post?.text ? post.text.substring(0, 140) : 'Discover, share, and vote on the best quotes. Join the Quote.Vote community!';
  const ogImage = post?.imageUrl || 'https://quote.vote/og-default.jpg';
  const ogUrl = post ? `https://quote.vote/post/${post._id}` : 'https://quote.vote/';

  // To reset the scroll when the selected post changes
  useEffect(() => {
    window.scrollTo(0, 0)
    if (post) {
      setPostHeight(document.getElementById('post')?.clientHeight)
    }
  }, [post])

  useEffect(() => {
    refetchPost({ postId: idSelector })
  }, [idSelector, refetchPost])

  let messageRoomId
  let title
  if (post) {
    messageRoomId = postData.post.messageRoom?._id
    title = postData.post.title
  }

  const {
    loading: loadingMessages,
    error: messageError,
    data: messageData,
    refetch: refetchMessages,
  } = useQuery(GET_ROOM_MESSAGES, {
    skip: !messageRoomId,
    variables: { messageRoomId },
  })

  // Add error logging
  if (messageError) {
    console.error('[CLIENT] PostPage - message query error:', messageError);
  }

  useSubscription(NEW_MESSAGE_SUBSCRIPTION, {
    skip: !messageRoomId,
    variables: { messageRoomId },
    onSubscriptionData: async () => {
      await refetchMessages()
    },
  })

  if (postError) return <Redirect to="/error" />

  const { messages } = (!loadingMessages && messageData) || []

  const { comments, votes, quotes } = post || {
    comments: [],
    votes: [],
    quotes: [],
  }

  // Filter out deleted comments
  const filteredComments = comments.filter((c) => !c.deleted)
  // Filter out deleted quotes
  const filteredQuotes = quotes.filter((q) => !q.deleted)
  // Filter out deleted votes
  const filteredVotes = votes.filter((v) => !v.deleted)

  const postActions = useMemo(() => {
    let postActions = []

    if (!isEmpty(filteredComments)) {
      postActions = postActions.concat(
        filteredComments.map((comment) => ({
          ...comment,
          __typename: 'Comment',
          commentQuote:
            comment.endWordIndex > comment.startWordIndex
              ? post.text
                  .substring(comment.startWordIndex, comment.endWordIndex)
                  .replace(/(\r\n|\n|\r)/gm, '')
              : null,
        }))
      )
    }

    if (!isEmpty(filteredVotes)) {
      postActions = postActions.concat(
        filteredVotes.map((vote) => ({
          ...vote,
          __typename: 'Vote',
        }))
      )
    }

    if (!isEmpty(filteredQuotes)) {
      postActions = postActions.concat(
        filteredQuotes.map((quote) => ({
          ...quote,
          __typename: 'Quote',
        }))
      )
    }

    // Add messages to postActions so they get displayed
    if (!isEmpty(messages)) {
      postActions = postActions.concat(
        messages.map((message) => ({
          ...message,
          __typename: 'Message',
          text: message.text, // Ensure text field is present for PostActionCard to recognize it as a message
        }))
      )
    }

    return postActions
  }, [comments, votes, quotes, messages])

  const { url } = (!loadingPost && post) || {}

  if (isMobile) {
    // Mobile layout - vertical stacking
    return (
      <>
        <Helmet>
          <title>{ogTitle}</title>
          <meta property="og:title" content={ogTitle} />
          <meta property="og:description" content={ogDescription} />
          <meta property="og:image" content={ogImage} />
          <meta property="og:url" content={ogUrl} />
          <meta property="og:type" content="article" />
          <meta name="twitter:card" content="summary_large_image" />
          <meta name="twitter:title" content={ogTitle} />
          <meta name="twitter:description" content={ogDescription} />
          <meta name="twitter:image" content={ogImage} />
        </Helmet>
        <div className={classes.mobileContainer}>
          <div className={classes.mobilePostSection} id="post">
            {loadingPost ? (
              <PostSkeleton />
            ) : (
              <Post
                post={post}
                loading={loadingPost}
                user={user}
                postHeight={postHeight}
                postActions={postActions}
                refetchPost={refetchPost}
              />
            )}
          </div>
          <div className={classes.mobileInteractionSection}>
            <div className={classes.mobileMessagesContainer}>
              <PostActionList
                loading={loadingPost}
                postActions={postActions}
                postUrl={url}
                refetchPost={refetchPost}
              />
            </div>
            <div className={classes.mobileChatInputContainer}>
              <PostChatSend messageRoomId={messageRoomId} title={title} />
            </div>
          </div>
        </div>
      </>
    )
  }

  // Desktop layout - side by side panels
  return (
    <>
      <Helmet>
        <title>{ogTitle}</title>
        <meta property="og:title" content={ogTitle} />
        <meta property="og:description" content={ogDescription} />
        <meta property="og:image" content={ogImage} />
        <meta property="og:url" content={ogUrl} />
        <meta property="og:type" content="article" />
        <meta name="twitter:card" content="summary_large_image" />
        <meta name="twitter:title" content={ogTitle} />
        <meta name="twitter:description" content={ogDescription} />
        <meta name="twitter:image" content={ogImage} />
      </Helmet>
      <div className={classes.desktopContainer}>
        {/* Left Panel - Post Content */}
        <div className={classes.desktopPostSection} id="post">
          {loadingPost ? (
            <PostSkeleton />
          ) : (
            <Post
              post={post}
              loading={loadingPost}
              user={user}
              postHeight={postHeight}
              postActions={postActions}
              refetchPost={refetchPost}
            />
          )}
        </div>
        {/* Right Panel - Actions, Chat Messages, and Chat Input */}
        <div className={classes.desktopInteractionSection}>
          <div className={classes.desktopMessagesContainer}>
            <PostActionList
              loading={loadingPost}
              postActions={postActions}
              postUrl={url}
              refetchPost={refetchPost}
            />
          </div>
          <div className={classes.desktopChatInputContainer}>
            <PostChatSend messageRoomId={messageRoomId} title={title} />
          </div>
        </div>
      </div>
    </>
  )
}

PostPage.propTypes = {
  postId: PropTypes.string,
}

export default PostPage
