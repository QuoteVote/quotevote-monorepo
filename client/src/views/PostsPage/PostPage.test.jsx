// Stub window.scrollTo for environments where setupTests may not yet run
if (typeof window !== 'undefined' && typeof window.scrollTo !== 'function') {
  // eslint-disable-next-line no-empty-function
  window.scrollTo = function () {}
}
import React from 'react'
import { render } from '@testing-library/react'
// Mock heavy children before importing the tested module to avoid side-effects
vi.mock('../../components/Post/Post', () => ({
  __esModule: true,
  default: () => <div data-testid="mock-post" />,
}))
vi.mock('../../components/PostActions/PostActionList', () => ({
  __esModule: true,
  default: () => <div data-testid="mock-postactions" />,
}))
vi.mock('../../components/PostChat/PostChatSend', () => ({
  __esModule: true,
  default: () => <div data-testid="mock-postchatsend" />,
}))

// Mock Apollo hooks used inside PostPage so tests don't attempt network or
// subscription behavior during simple render checks.
vi.mock('@apollo/react-hooks', async () => {
  const actual = await vi.importActual('@apollo/react-hooks')
  return {
    ...actual,
    useQuery: () => ({ loading: false, data: { post: null }, refetch: () => {} }),
    useSubscription: () => ({}),
  }
})

import PostPage from './PostPage'
import withTestWrapper from '../../hoc/withTestWrapper'

const PostPageWrapper = withTestWrapper(PostPage)

describe('Post Page test -', () => {
  it('renders mocked children', () => {
    const { getByTestId } = render(
      <PostPageWrapper />,
    )
    // assert our mocked children are rendered
    expect(getByTestId('mock-post')).toBeTruthy()
    expect(getByTestId('mock-postactions')).toBeTruthy()
    expect(getByTestId('mock-postchatsend')).toBeTruthy()
  })
})
