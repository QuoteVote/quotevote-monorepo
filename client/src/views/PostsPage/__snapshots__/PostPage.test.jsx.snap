// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Post Page test - > renders correctly 1`] = `
<div
  class="makeStyles-desktopContainer-3"
>
  <div
    class="makeStyles-desktopPostSection-5"
    id="post"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-6fm6rz-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-ku7d13-MuiTypography-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row css-wjsns7-MuiGrid-root"
              justify="flex-start"
            >
              <div
                class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
                  style="width: 200px; margin-right: 10px;"
                />
              </div>
              <div
                class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
                  style="width: 30px; height: 30px;"
                />
              </div>
              <div
                class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
                  style="width: 30px; height: 30px;"
                />
              </div>
            </div>
          </span>
        </div>
        <div
          class="MuiCardHeader-action css-116l9hl-MuiCardHeader-action"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row css-wjsns7-MuiGrid-root"
            justify="flex-start"
            style="margin-top: 15px;"
          >
            <div
              class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
                style="width: 30px; margin-right: 5px;"
              />
            </div>
            <div
              class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
            >
               
              /
               
            </div>
            <div
              class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
                style="width: 30px; margin-left: 5px;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
        style="padding: 0px; margin-left: 15px;"
      >
        <div
          class="MuiCardHeader-avatar css-1r9wl67-MuiCardHeader-avatar"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
            style="width: 50px; height: 50px;"
          />
        </div>
        <div
          class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-1nmqws6-MuiTypography-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
              style="width: 80px; margin-bottom: 5px;"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-ihhumc-MuiTypography-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
              style="width: 130px;"
            />
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1k2ey29-MuiSkeleton-root"
          style="height: 30px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1k2ey29-MuiSkeleton-root"
          style="height: 30px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1k2ey29-MuiSkeleton-root"
          style="height: 30px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-1k2ey29-MuiSkeleton-root"
          style="width: 50%; height: 30px;"
        />
      </div>
      <div
        class="MuiCardActions-root css-1wwmo7x-MuiCardActions-root"
        style="margin-left: 5px;"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
          style="width: 130px; height: 30px; margin-right: 5px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
          style="width: 130px; height: 30px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
          style="width: 30px; height: 30px; margin-left: auto;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
          style="width: 30px; height: 30px; margin-right: 5px; margin-left: 5px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
          style="width: 30px; height: 30px;"
        />
      </div>
    </div>
  </div>
  <div
    class="makeStyles-desktopInteractionSection-9"
  >
    <div
      class="makeStyles-desktopMessagesContainer-10"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row css-1ngyxha-MuiGrid-root"
        justify="space-between"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
        style="height: 118px;"
      />
      <br />
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
        style="height: 118px;"
      />
      <br />
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse css-1lblo9c-MuiSkeleton-root"
        style="height: 118px;"
      />
    </div>
    <div
      class="makeStyles-desktopChatInputContainer-11"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row makeStyles-root-15 css-wjsns7-MuiGrid-root"
        justify="center"
      >
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 makeStyles-chat-16 css-1edfpdg-MuiTypography-root"
          >
            Chat
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-1dtoofl-MuiPaper-root"
          >
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary makeStyles-input-17 css-1emi5m7-MuiInputBase-root"
            >
              <input
                class="MuiInputBase-input css-1jrstcx-MuiInputBase-input"
                placeholder="type a message..."
                type="text"
                value=""
              />
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium makeStyles-send-18 css-1rjloaz-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <img
                alt="send"
                src="/assets/SendIcon.svg"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
